### PR TITLE
Add a revive method

### PR DESCRIPTION
**Pull request by stuartskelton**
Now with added revival. I still need to add some docs. but I think it works nicely, this builds off the load process code, and will rebase once the previous PR is accepted.

### CODE REVIEW

File: lib/Schedule/LongSteps.pm [hunk 1]
```diff
@@ -455,4 +469,102 @@ sub find_process{
...
+    }
+
+    return $loaded_process;
+}
+
+sub revive {
+    my ( $self, $process_id, $revive_to ) = @_;
+
+    # find process
+    my $stored_process = eval { $self->find_process($process_id) };
```
Comment (jeteve) [inline lib/Schedule/LongSteps.pm]: Crash if not there.

File: lib/Schedule/LongSteps.pm [hunk 2]
```diff
@@ -455,4 +469,102 @@ sub find_process{
...
+    if ($@) {
+        $log->critical("$process_id does not exist");
+        return;
+    }
+
+    # check to see if its allowed to be revived.
+    # has it stopped
+    if ( $stored_process->status() ne "terminated" ) {
+        $log->critical("$process_id does not have a status of 'terminated'");
```
Comment (jeteve) [inline lib/Schedule/LongSteps.pm]: crash

File: lib/Schedule/LongSteps.pm [hunk 3]
```diff
@@ -455,4 +469,102 @@ sub find_process{
...
+        return;
+    }
+
+    # load the process and check if process have the method to revive_to
+    # if revive $revive_to was not passed, used the function we failed on.
+    # and check that also, just in case we attempt to revive on a method
+    # that was previously removed.
+    $revive_to = $stored_process->what() unless $revive_to;
+    my $loaded_process = eval { $self->_load_process($stored_process) };
```
Comment (jeteve) [inline lib/Schedule/LongSteps.pm]: crash

File: lib/Schedule/LongSteps.pm [hunk 4]
```diff
@@ -455,4 +469,102 @@ sub find_process{
...
+
+    if ($@) {
+        $log->critical("Failed to load the process: $@");
+        return;
+    }
+
+    unless ($loaded_process->can($revive_to)) {
```
Comment (jeteve) [inline lib/Schedule/LongSteps.pm]: crash as well.

File: t/revive_with_revival_step.t [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                what   => 'do_break_stuff',
+                run_at => DateTime->now(),
+                state =>
+                  { %{ $self->state() }, n_tries => 10, sandwich => 'toast' }
+            }
+        );
+    }
+
+    sub revive_do_break_stuff {
```
Comment (jeteve) [inline t/revive_with_revival_step.t:33]: revive should return $self->new_step( ... );

File: lib/Schedule/LongSteps.pm [hunk 6]
```diff
@@ -455,4 +469,102 @@ sub find_process{
...
+        $log->critical("Unable revive $process_id to $revive_to");
+        return;
+    }
+
+    my $revive_method = "revive_$revive_to";
+    if ( $loaded_process->can($revive_method) ) {
+        $log->info("Running $revive_method function");
```
Comment (jeteve) [inline lib/Schedule/LongSteps.pm]: $stored_process->what( $revive_method );
$stored_process->run_at( now() );
$stored_process->error(undef);
$stored_process->update();

File: lib/Schedule/LongSteps.pm [hunk 7]
```diff
@@ -455,4 +469,102 @@ sub find_process{
...
+        $log->critical("Failed to load the process: $@");
+        return;
+    }
+
+    unless ($loaded_process->can($revive_to)) {
+        $log->critical("Unable revive $process_id to $revive_to");
+        return;
+    }
+
+    my $revive_method = "revive_$revive_to";
```
Comment (jeteve) [inline lib/Schedule/LongSteps.pm]: don't need that.

What you can do is on the process base class:

An instance method that returns all the 'revive_*' methods.  (overridable in sub processes).

File: lib/Schedule/LongSteps.pm [hunk 8]
```diff
@@ -476,6 +484,50 @@ sub load_process {
...
+    # that was previously removed.
+    my $loaded_process = $self->_load_stored_process($stored_process);
+
+    $revive_to = $stored_process->what() unless $revive_to;
+    unless ( $loaded_process->can($revive_to) ) {
+        confess "Unable revive $process_id to $revive_to";
+    }
+
+    # set the next step, if there is a revive_$revive_to,this takes precidence
+    my $revive_method = "revive_$revive_to";
```
Comment (jeteve) [inline lib/Schedule/LongSteps.pm]: No magic prefixing please. Plus it's inconsistent with the previous defaulting to 'what'.

File: lib/Schedule/LongSteps.pm [hunk 9]
```diff
@@ -476,6 +484,50 @@ sub load_process {
...
+    # set the revivie point
+    $stored_process->what($revive_to);
+
+    # remove the any error
+    $stored_process->error(undef);
+
+    # change the status to paused
+    $stored_process->status("paused");
+
+    # reschedual the process
```
Comment (jeteve) [inline lib/Schedule/LongSteps.pm]: typo

File: lib/Schedule/LongSteps.pm [hunk 10]
```diff
@@ -476,6 +484,50 @@ sub load_process {
...
+    confess("$process_id does not have a status of 'terminated'") if ( $stored_process->status() ne "terminated" );
+
+    # load the process and check if process have the method to revive_to
+    # if revive $revive_to was not passed, used the function we failed on.
+    # and check that also, just in case we attempt to revive on a method
+    # that was previously removed.
+    my $loaded_process = $self->_load_stored_process($stored_process);
+
+    $revive_to = $stored_process->what() unless $revive_to;
+    unless ( $loaded_process->can($revive_to) ) {
```
Comment (jeteve) [inline lib/Schedule/LongSteps.pm]: separate those two things. The spacing implies they are logically linked together. Which is not the case. Confuses me.

File: lib/Schedule/LongSteps.pm [hunk 11]
```diff
@@ -476,6 +484,50 @@ sub load_process {
...
+
+    # remove the any error
+    $stored_process->error(undef);
+
+    # change the status to paused
+    $stored_process->status("paused");
+
+    # reschedual the process
+    $stored_process->run_at(DateTime->now());
+    $stored_process->update();
```
Comment (jeteve) [inline lib/Schedule/LongSteps.pm]: Too many redundant comments. Less blank lines would show all those things live together.

Comment (stuartskelton) [reply lib/Schedule/LongSteps.pm]: done

File: lib/Schedule/LongSteps.pm (hunk 10 above)
Comment (stuartskelton) [reply lib/Schedule/LongSteps.pm]: sorted

File: lib/Schedule/LongSteps.pm (hunk 8 above)
Comment (stuartskelton) [reply lib/Schedule/LongSteps.pm]: yeah, it would be confusing, if you did revive(do_xyz) and it redirects it to 'revive_do_xyz'

**Pull request merged**